src/nlp_expr.jl

```diff
@@ -0,0 +1,1063 @@
+    f::Function,
+    args::Vararg{Function,N};
+    name::Symbol = Symbol(f),
+) where {N}
+    nargs = 1 + N
+    if !(1 <= nargs <= 3)
+        error(
+            "Unable to register operator $name: invalid number of functions " *
+            "provided. Got $nargs, but expected 1 (if function only), 2 (if " *
+            "function and gradient), or 3 (if function, gradient, and " *
+            "hesssian provided)",
+        )
+    end
+    # TODO(odow): we could add other checks here, but we won't for now because
+    # down-stream solvers in MOI can add their own checks, and any solver using
+    # MOI.Nonlinear will automatically check for autodiff and common mistakes
+    # and throw a nice informative error.
+    MOI.set(model, MOI.UserDefinedFunction(name, dim), tuple(f, args...))
+    return NonlinearOperator(name, f)
+end
+
+function _redefinition_constant_error(op)
+    return """
+    Unable to register the nonlinear operator `:$op` because it has the same
+    symbolic identifier as an existing object in the current scope.
+
+    For example, this code will error:
+    ```julia
+    model = Model()
+    f(x) = x^2
+    @register(model, f, 1, f)
+    ```
+    because it is equivalent to:
+    ```julia
+    model = Model()
+    f(x) = x^2
+    f = register_nonlinear_operator(model, 1, f; name = :f)
+    ```
+
+    To fix, use a unique name, like `op_$op`:
+    ```julia
+    model = Model()
+    f(x) = x^2
+    @register(model, op_f, 1, f)
+    @expression(model, op_f(x))
+    ```
+    """
+end
+
+"""
+    @register(model, operator, dim, f[, ∇f[, ∇²f]])
+
+Register the nonlinear operator `operator` in `model`, and create a new
+[`NonlinearOperator`](@ref) object called `operator` in the current scope.
+
+## Example
+
+```jldoctest
+julia> model = Model();
+
+julia> @variable(model, x)
+x
+
+julia> f(x::Float64) = x^2
+f (generic function with 1 method)
+
+julia> ∇f(x::Float64) = 2 * x
+∇f (generic function with 1 method)
+
+julia> ∇²f(x::Float64) = 2.0
+∇²f (generic function with 1 method)
+
+julia> @register(model, op_f, 1, f, ∇f, ∇²f)
+NonlinearOperator(:op_f, f)
+
+julia> @objective(model, Min, op_f(x))
+op_f(x)
+
+julia> op_f(2.0)
+4.0
+
+julia> model[:op_f]
+NonlinearOperator(:op_f, f)
+
+julia> model[:op_f](x)
+op_f(x)
+```
+
+## Non-macro version
+
+This macro is provided as helpful syntax that matches the style of the rest of
+the JuMP macros. However, you may also register operators outside the macro
+using [`register_nonlinear_operator`](@ref). For example:
+
+```jldoctest
+julia> model = Model();
+
+julia> f(x) = x^2
+f (generic function with 1 method)
+
+julia> @register(model, op_f, 1, f)
+NonlinearOperator(:op_f, f)
+```
+is equivalent to
+```jldoctest
+julia> model = Model();
+
+julia> f(x) = x^2
+f (generic function with 1 method)
+
+julia> op_f = model[:op_f] = register_nonlinear_operator(model, 1, f; name = :op_f)
+NonlinearOperator(:op_f, f)
+```
+"""
+macro register(model, op, dim, f, args...)
+    if op == f
```

Comment:
It won't work if you do
```julia
f() = 1
a = f
@register(model, f, 0, a)
```
You need to call a function with the values of `op` and `f` and then in the function you check if `op == Symbol(f)`